Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1422 @@
+          .init(
+            code: Status.Code(httpStatusCode: httpStatusCode),
+            message: "Unexpected non-200 HTTP Status Code."
+          )
+        )
+      }
+
+      let contentTypeHeader = metadata.first(name: GRPCHTTP2Keys.contentType.rawValue)
+      guard contentTypeHeader.flatMap(ContentType.init) != nil else {
+        return .failedRequest(
+          .init(code: .internalError, message: "Missing \(GRPCHTTP2Keys.contentType) header")
+        )
+      }
+
+      if endStream {
+        // This is a trailers-only response.
+        self.state = .clientClosedServerClosed(.init(previousState: state))
+      } else {
+        var inboundEncoding: CompressionAlgorithm?
+        if let serverEncoding = metadata.first(name: GRPCHTTP2Keys.encoding.rawValue) {
+          guard let parsedEncoding = CompressionAlgorithm(rawValue: serverEncoding) else {
+            return .failedRequest(
+              .init(
+                code: .internalError,
+                message: "The server picked a compression algorithm the client does not know about."
+              )
+            )
+          }
+          inboundEncoding = parsedEncoding
+        }
+
+        self.state = .clientClosedServerOpen(
+          .init(
+            previousState: state,
+            decompressionAlgorithm: inboundEncoding
+          )
+        )
+      }
+      return .receivedMetadata(Metadata(headers: metadata))
+    case .clientClosedServerOpen(let state):
+      if endStream {
+        self.state = .clientClosedServerClosed(.init(previousState: state))
+      } else {
+        // This state is valid: server can send trailing metadata without grpc-status
+        // or END_STREAM set, and follow it with an empty message frame where they are set.
+        ()
+      }
+      return .receivedMetadata(Metadata(headers: metadata))
+    case .clientClosedServerClosed:
+      // We could end up here if we received a grpc-status header in a previous
+      // frame (which would have already close the server) and then we receive
+      // an empty frame with EOS set.
+      // We wouldn't want to throw in that scenario, so we just ignore it.
+      // Note that we don't want to ignore it if EOS is not set here though, as
+      // then it would be an invalid payload.
+      if !endStream || metadata.count > 0 {
+        throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+          "Server is closed, nothing could have been sent."
+        )
+      }
+      return .receivedMetadata([])
+    case .clientIdleServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server cannot have sent metadata if the client is idle."
+      )
+    case .clientOpenServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server is closed, nothing could have been sent."
+      )
+    }
+  }
+
+  private mutating func clientReceive(bytes: ByteBuffer, endStream: Bool) throws {
```

Comment:
Why do some methods operate on `[UInt8]` and others on `ByteBuffer`?

Reply:
When we receive data, we receive `ByteBuffer`s from the channel. However we then deal with `[UInt8]` internally, as we decode bytes into RPC messages, meaning the outbound methods operate on `[UInt8]`.